YouTube Shorts now shows dislike counts

RyannDaGreat opened the pull request:
The current version of this extension doesn't work with YouTube Shorts. It shows this:

<img width="1440" alt="Screen Shot 2022-03-18 at 5 51 47 PM" src="https://user-images.githubusercontent.com/17944835/159088698-a3ab9e73-7e2c-421a-8227-9fa2a43f03e1.png">

This commit allows dislike counts to be shown in YouTube Shorts. With this change, it now looks like this:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/17944835/159088981-d73b3f10-4f05-4334-849a-5c34a8828184.png">

Note: It is important that we add this feature, becuase YouTube is now forcibly converting short videos into YouTube shorts. Many videos that are under a minute long on my channel have been turned into YouTube Short videos (such as [this video](https://www.youtube.com/shorts/eDbfVg7R_JA)). There's no option to turn that off in the creator studio, which means many videos that used to display dislikes with this plugin no longer do (without this PR).

This pull request addresses two issues:
https://github.com/Anarios/return-youtube-dislike/issues/382
https://github.com/Anarios/return-youtube-dislike/issues/484

avi12 commented on Extensions/combined/src/state.js:
```diff
@@ -27,6 +27,10 @@ function isMobile() {
 }
 
+function isShorts() {
+  return new URL(document.URL).pathname.startsWith("/shorts/")
```
Why not simply:
```js
return location.pathname.startsWith("/shorts");
```

RyannDaGreat replied:
Done!

avi12 commented on Extensions/combined/src/utils.js:
```diff
@@ -45,10 +45,24 @@ function getVideoId(url) {
+function isInViewport(element) {
+  //Code from: https://stackoverflow.com/questions/19669786/check-if-element-is-visible-in-dom
+  const rect = element.getBoundingClientRect();
+  return (
+      rect.top >= 0 &&
+      rect.left >= 0 &&
+      rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
+      rect.right <= (window.innerWidth || document.documentElement.clientWidth)
```
I personally prefer having this sort of stuff clean:
```js
const height = innerHeight || document.documentElement.clientHeight;
const width = innerWidth || document.documentElement.clientWidth;
return (
  rect.top >= 0 &&
  rect.left >= 0 &&
  rect.bottom <= height &&
  rect.right <= width
);
```

RyannDaGreat replied:
Done!

avi12 commented on Extensions/combined/src/utils.js:
```diff
@@ -45,10 +45,24 @@ function getVideoId(url) {
+}
+
 function isVideoLoaded() {
   const videoId = getVideoId(window.location.href);
```
`window.` could be removed

RyannDaGreat replied:
Done!

avi12 replied:
I meant in the `window.location.href` part

RyannDaGreat replied:
Oh, that's out of the scope of this PR (the code I added doesn't directly use that function)

RyannDaGreat commented:
@avi12 Hey! I made your requested changes and tested it - it still works :)

avi12 commented:
@RyannDaGreat I'm not a contributor in this repository, perhaps one should review your code
I was just offering my knowledge :)

avi12 commented on Extensions/combined/src/utils.js:
```diff
@@ -45,10 +45,26 @@ function getVideoId(url) {
 }
 
+function isInViewport(element) {
+  //Code from: https://stackoverflow.com/questions/19669786/check-if-element-is-visible-in-dom
```
I don't think comments are necessary for the code
Rather, it could be inside a PR comment

RyannDaGreat replied:
Personally, I like to keep track of where I learn how to do things - that way, if the browser api changes in the future, there might be an updated answer on that link on stackoverflow

avi12 replied:
You should link to the answer that provided the function, not to the question

RyannDaGreat commented:
> @RyannDaGreat I'm not a contributor in this repository, perhaps one should review your code I was just offering my knowledge :)

Any idea how to get their attention? Lol

avi12 commented:
You should assign 6 of the contributors to be reviewers
![image](https://user-images.githubusercontent.com/6422804/159096557-b319a927-e46c-4e0d-816c-428794809c97.png)

RyannDaGreat commented:
> You should assign 6 of the contributors to be reviewers ![image](https://user-images.githubusercontent.com/6422804/159096557-b319a927-e46c-4e0d-816c-428794809c97.png)

I've never done that before...do you know how? That 
```
Assignees
No one assigned
```
text isn't clickable

avi12 commented:
I see, then you can just mention 6 of them with @

> I've never done that before...do you know how? That

To be fair, I don't have a huge experience with code reviews either
I mostly gained my code reviewing experience through the company I work for

RyannDaGreat commented:
I see, I'll message the top 6 contributors here:

@Anarios 
@Frontesque 
@PickleNik 
@DARKDRAGON532 
@sy-b 
@leonbubova 

Hey guys, could you please review this PR?

RyannDaGreat commented:
I just hope this gets reviewed...I see some other pending PR's that are nearly a month old...

DARKDRAGON532 reviewed (approved):
only anarios can merge

sy-b commented:
Well done 

Here's a checklist for this feature

- Extension
  - Desktop
    - [x] Show Dislikes
    - [ ] Send Dislike
  
  - Mobile
    - [ ] Show Dislikes
    - [ ] Send Dislike
   
- Userscript

  - Desktop
    - [ ] Show Dislikes
    - [ ] Send Dislike
  
  - Mobile
    - [ ] Show Dislikes
    - [ ] Send Dislike

sy-b reviewed (changes requested):
Well done 

Here's a checklist for this feature

- Extension
  - Desktop
    - [x] Show Dislikes
    - [ ] Send Dislike
    - [ ] Send Like
  
  - Mobile
    - [ ] Show Dislikes
    - [ ] Send Dislike
    - [ ] Send Like


- Userscript

  - Desktop
    - [ ] Show Dislikes
    - [ ] Send Dislike
    - [ ] Send Like
  
  - Mobile
    - [ ] Show Dislikes
    - [ ] Send Dislike
    - [ ] Send Like
 
 ---
 
 I know that its hard to implement this without bloating or restructuring.
 I am in favor of restructuring things to avoid mess.
 Good luck 🙂

RyannDaGreat commented on Extensions/combined/ryd.content-script.js:
```diff
@@ -35,7 +36,7 @@ let jsInitChecktimer = null;
 
 function setEventListeners(evt) {
   function checkForJS_Finish() {
-    if (getButtons()?.offsetParent && isVideoLoaded()) {
```
Did you test this to make sure this works? I added the "isShorts() ||" because without it, it wouldn't work

RyannDaGreat commented on Extensions/combined/src/bar.js:
```diff
@@ -1,5 +1,5 @@
 import { getButtons } from "./buttons";
-import { likesDisabledState } from "./state";
```
isMobile is referenced later in this file, and threw an annoying amount of javascript errors in the console when it wasn't imported. Perhaps it should be a separate PR? But yeah, this is a missing import...it's a bug; isMobile should be imported because of a later line in this file

avi12 commented on Extensions/UserScript/Return Youtube Dislike.user.js:
```diff
@@ -30,13 +30,34 @@ let likesvalue = 0;
 let dislikesvalue = 0;
 
 let isMobile = location.hostname == "m.youtube.com";
+let isShorts = location.pathname.startsWith("/shorts")
```
I don't know how a UserScript works, but I believe it should be a function, such as:
```js
const isShorts = () => location.pathname.startsWith("/shorts")
```
because otherwise, the value will remain the same even after the user navigates to a non-Shorts video

RyannDaGreat replied:
Oh mah lord - you replied so fast XD (Aweome lmao, I didn't even check this page yet)
And that's a good point - I'll add that

RyannDaGreat replied:
I'm still going to use 'let' instead of 'const' because that seems to be what the other lines around it are doing lol (makes the code more ａｅｓｔｈｅｔｉｃ)

avi12 replied:
> Oh mah lord - you replied so fast XD (Aweome lmao, I didn't even check this page yet) And that's a good point - I'll add that

Lmao it's the opposite
I came to see the code and you just happened to have pushed a commit 4 minutes ago by the time I reviewed the code

avi12 commented on Extensions/UserScript/Return Youtube Dislike.user.js:
```diff
@@ -30,13 +30,34 @@ let likesvalue = 0;
 }
 
+function isInViewport(element) {
+  const rect = element.getBoundingClientRect();
+  const height = innerHeight || document.documentElement.clientHeight;
+  const width = innerWidth || document.documentElement.clientWidth;
```
According to what you said, these should also be `let`

RyannDaGreat replied:
I see where you're coming from...but not really. It's because if it was 'const', then 'isShorts' and 'isMobile' wouldn't be aligned any more. It's because isShorts was a function in the Extension source code that was directly below the isMobile function that I decided to put the isShorts variable below the isMobile variable, so I used the same keyword. I did this to maintain as much consistency as I can between the userscript and the extension codebase.

For the same reason, I want to keep isInViewport as similar as possible to the original code base - meaning keeping the same source code verbatim because that works. So, I'm sticking with const.

avi12 commented on Extensions/combined/src/state.js:
```diff
@@ -27,6 +27,10 @@ function isMobile() {
   return location.hostname == "m.youtube.com";
 }
 
+function isShorts() {
+  return location.pathname.startsWith("/shorts")
+}
```
Why is this code duplicated?

RyannDaGreat replied:
Duplicated from where?

avi12 replied:
https://github.com/Anarios/return-youtube-dislike/pull/511/files#diff-567d3ac937cc83968cffa40f0c919c000d27d84fcb9836e7c4cbf6ba81e9eaddR33

This violates the DRY concept

RyannDaGreat replied:
I agree...it does. You're comparing `Return Youtube Dislike.user.js` to `state.js` 
Unfortunately, there's no getting around this. A lot of the code (nearly all of the functions, in fact) in `Return Youtube Dislike.user.js` is duplicate code - not just my stuff.

RyannDaGreat replied:
Code in `Return Youtube Dislike.user.js` can't access `state.js` because it's just copy-pasted from `Return Youtube Dislike.user.js` into tampermonkey etc.

RyannDaGreat replied:
Meanwhile, code in `state.js` is meant for extensions - not UserScripts and so it doesn't access any of the UserScript code. Perhaps it should! It would mean less redundancy! I didn't make this repo tho lol

RyannDaGreat commented:
> Well done
> 
> Here's a checklist for this feature
> 
> * Extension
>   
>   * Desktop
>     
>     * [x]  Show Dislikes
>     * [ ]  Send Dislike
>     * [ ]  Send Like
>   * Mobile
>     
>     * [ ]  Show Dislikes
>     * [ ]  Send Dislike
>     * [ ]  Send Like
> * Userscript
>   
>   * Desktop
>     
>     * [ ]  Show Dislikes
>     * [ ]  Send Dislike
>     * [ ]  Send Like
>   * Mobile
>     
>     * [ ]  Show Dislikes
>     * [ ]  Send Dislike
>     * [ ]  Send Like
> 
> I know that its hard to implement this without bloating or restructuring. I am in favor of restructuring things to avoid mess. Good luck 🙂

Hey @sy-b ! So, I'm not sure how to test mobile...I have an iPhone, and I don't want to Jailbreak it. Also, YouTube Vanced has been shut down due to legal issues: [(The Verge)](https://www.theverge.com/2022/3/13/22975890/youtube-vanced-app-discontinued-shutting-down-legal-reasons)
Even if it wasn't shut down, I don't know how I would test it without modifying their app...which I don't think I have control over...how would one test it?
That said, I suspect it will probably work on mobile already. Unlike the regular youtube site, the desktop YouTube Shorts pages look super similar to the mobile ones - so I bet they use the same HTML tags (which is the main difference between the mobile and desktop code). I've added one change to getButtons() in buttons.js, which should let it work on mobile devices (but again, somebody please test this. The UserScript support I added already has this change.

Also, as for sending likes and dislikes: it appears to work already! Here's what I did to test this: 
The sendVote(), likeClicked() and dislikeClicked() functions fire correctly when a like or dislike button is clicked in YouTube shorts. When a dislike button is clicked, for example, the sendVote() function gets sent a -1 value. I tested this by inserting alerts and console logs into these functions and then using the like and dislike buttons on YouTube shorts. This also worked with the UserScript. (But again, I didn't test this on Mobile)

avi12 commented:
> Hey @sy-b ! So, I'm not sure how to test mobile...I have an iPhone, and I don't want to Jailbreak it. 

You can pull up an Android emulator that has browsers that support installing Return YouTube Dislike (either as an extension or a UserScript), e.g. Firefox

RyannDaGreat commented:
How do I install an extension from source though? I can't use the one from the store, as that wouldn't have my code. Do you have any emulators in mind?
Also, do you have any idea how we'd test for IOS?

RyannDaGreat commented:
Also, @sy-b @avi12 I was thinking of reordering and squashing my four commits down to two (so there are less commits once this is merged). Should I go ahead with it, or leave it as four separate commits? (The last commit, for example, is a very small change)

The two commits would be:
<img width="479" alt="image" src="https://user-images.githubusercontent.com/17944835/159156168-b93584bb-afaa-4c51-bd2a-4635ef183168.png">

Right now, the four commits are:
<img width="651" alt="image" src="https://user-images.githubusercontent.com/17944835/159156182-c00a7130-8fbf-45f3-a6ab-26b0658264ac.png">

avi12 commented:
> (The last commit, for example, is a very small change)

Please take a look at [amend](https://www.atlassian.com/git/tutorials/rewriting-history)-ing, a Git feature (i.e. not specific to GitHub)

avi12 commented:
> @avi12 How do I install an extension from source though?

When it comes to Firefox, I don't know TBH; the only source I have is https://extensionworkshop.com/documentation/develop/developing-extensions-for-firefox-for-android

RyannDaGreat commented:
For now, let's keep the scope of this PR to desktop browsers - for the sake of getting the feature out as soon as possible. Testing on mobile seems difficult, and I don't think any of my code makes it more difficult to implement one for a mobile device (the code changes are very small). That can be for a future PR.

avi12 commented:
I saw in the `Docs/` folder [this one](https://github.com/Anarios/return-youtube-dislike/wiki/Downloading,-Installing-&-Using#mobile)

RyannDaGreat commented:
Ok - I successfully added support for and tested both extension and UserScript support for mobile! I tested it by getting my computer to display YouTube's mobile website, [using a tutorial on reddit.](https://www.reddit.com/r/youtube/comments/6zn2pq/how_to_open_mobile_version_of_youtube_on_pc/)

_Chrome > Inspector > Responsive Design Mode
https://i.imgur.com/OsQUSo4.png
upd.: works best with this link — http://m.youtube.com/?persist_app=1&app=m and Show device frame option


I clicked that link and I can not get back to yotube desktop website


https://www.youtube.com/?app=desktop&persist_app=0 will change it back to desktop_

sy-b commented:
> I'm not sure how to test mobile

`Dev Tools` --> `Ctrl + Shift + M` or the mobile shaped button

RyannDaGreat commented:
> > I'm not sure how to test mobile
> 
> `Dev Tools` --> `Ctrl + Shift + M` or the mobile shaped button

Its ok, I figured it out - see the above comment

sy-b commented:
> (Update: I went ahead and did it. If anybody wants me to revert to 4 commits, let me know!)

`git reset --hard HEAD~4`

RyannDaGreat commented:
> > (Update: I went ahead and did it. If anybody wants me to revert to 4 commits, let me know!)
> 
> `git reset --hard HEAD~4`

@sy-b At this point I'm going to keep it the way it is lol - I've added two more commits to it

RyannDaGreat commented:
So, I'm guessing the next step is to get @Anarios on board with this?

sy-b commented:
> So, I'm guessing the next step is to get @Anarios on board with this?

He's a busy bee 🙂

RyannDaGreat commented:
Lol what - Ukraine isn't war enough?

Do you think I should try contacting him on discord?

sy-b commented:
> Lol what - Ukraine isn't war enough? (What do you mean?)

Didn't get that (are you using a translator or something?)

<br>

> Do you think I should try contacting him on discord?

No one is stopping you.

sy-b commented:
Some errors in `chrome://extensions/`.

sy-b reviewed (commented):
Issues
- Errors in `chrome://extensions/`. 
- Not working on mobile

sy-b commented:
add/edit 
fixes https://github.com/Anarios/return-youtube-dislike/issues/382
fixes https://github.com/Anarios/return-youtube-dislike/issues/484

Anarios commented:
Wow, that's a lot of work done, looking at it right now, thanks

Anarios reviewed (approved):
Great work, thanks!

Anarios commented:
Releasing this to beta today

sy-b commented:
![ryd gh_pull_484 errors 01](https://user-images.githubusercontent.com/94835959/159191517-b2f16a85-c056-4922-8705-b3425f57bf72.png)

sy-b commented:
I don't think this error is due to this PR.

avi12 commented:
> ![ryd gh_pull_484 errors 01](https://user-images.githubusercontent.com/94835959/159191517-b2f16a85-c056-4922-8705-b3425f57bf72.png)
> 


I honestly don't understand why is this snippet not
```js
document.querySelector("#menu-container, ytm-slim-video-action-bar-renderer").injectAdjacentHTML(...)
```
